### PR TITLE
OOP School Library Milestone #3: Associations Set-up

### PR DESCRIPTION
**Pull request by karlavdelgadof**
## OOP School Library Milestone #3: Associations Set-up :hammer_and_wrench: :school: :memo: :arrows_counterclockwise: 

These are the proposed changes :recycle:  for this project :arrow_down: :

- Created class `Classroom` :school:  with the following:
  - `@label` instance variable, initialized in the constructor.
  - Setter and getter for `@label` (attr_accessor).
- Created the `has-many/belongs-to` relationship between `Classroom` and `Student` as specified in the [requirements](https://github.com/microverseinc/curriculum-ruby/blob/main/oop/school_library_associations.md#:~:text=Create%20the%20many,set%20in%20it.). :heavy_check_mark: 
- Created class `Book` :book:  with the following:
   - `@title` and `@author` instance variables, initialized in the constructor.
   - Setters and getters for instance variables (attr_accessor).
- Created class `Rental` :dollar: with the following:
   - `@date` instance variable, initialized in the constructor.
   -  Setter and getter for @date (attr_accessor).
- Created the `many-to-many` (also has-many-through) relationship between `Person` and `Book` using the intermediate class `Rental` as specified in the [requirements](https://github.com/microverseinc/curriculum-ruby/blob/main/oop/school_library_associations.md#:~:text=Create%20the%20many,set%20in%20it.). :heavy_check_mark:

### CODE REVIEW

File: book.rb [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+  def initialize(title, author)
+    @title = title
+    @author = author
+    @rentals = []
+  end
+
+  def create_rental(rental)
+    @rentals << rental
+    rental.book = self
+  end
```
Comment (Reem-lab) [inline book.rb]: - [ ] Great job but I think here when creating the rental you should pass the person and date as a paramter to the function then you can generate a new rental. 🚀🔥💯

Something like this
> def add_rental(person, date)
    Rental.new(date, self, person)
  end

File: person.rb [hunk 2]
```diff
@@ -23,6 +24,11 @@ def correct_name
     @name
   end
 
+  def create_rental(rental)
+    @rentals << rental
+    rental.person = self
+  end
```
Comment (Reem-lab) [inline person.rb]: - [ ] Great job but I think here when creating the rental you should pass the book and date as a paramter to the function then you can generate a new rental. 🚀🔥💯

Something like this
> def add_rental(book, date)
    Rental.new(date,  book, slef)
  end

File: student.rb [hunk 3]
```diff
@@ -1,16 +1,19 @@
 require_relative 'person'
 
 class Student < Person
+  attr_accessor :classroom
```
Comment (Reem-lab) [inline student.rb]: - [ ] When setting the classroom for a student it also adds it but just as a `attr_reader` not as attr_accessor as required

image

**Review by Reem-lab — CHANGES_REQUESTED** [review]
# Hi @karlavdelgadof, 👋🏻👋🏻 

Good job so far! 💪🏻🚀🔥🔥💯💯
There are some issues that you still need to work on to go to the next project but you are almost there!


## Highlights
📌 No Linters error ✅
📌 Descriptive pull request✅
📌 Great working on the functionality✅
📌 Good commit messages ✅
📌 Professional README ✅

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: person.rb [hunk 4]
```diff
@@ -23,17 +24,13 @@ def correct_name
     @name
   end
 
+  def create_rental(_rental)
+    Rental.new(date, self, person)
+  end
```
Comment (HAMMAS-SALEEM) [inline person.rb]: - [x] The previous reviewer asked you to create your create_rental method something like this :arrow_down: cause you are not passing book and date as an argument in the function :blush:

```
def add_rental(book, date)
  Rental.new(date, book, slef)
end
```

**Review by HAMMAS-SALEEM — CHANGES_REQUESTED** [review]
# Required Changes :recycle:
almost-there

## Hey @karlavdelgadof :wave:

## Highlights :1st_place_medal:

:heavy_check_mark: Following GitFlow
:heavy_check_mark: Following some requirements
:heavy_check_mark: Meaningful commit messages
:heavy_check_mark: Descriptive PR Title

Good job on implementing some of the requested changes, there are some issues that you still need to work on to go to the next project but you are almost there! :blush:

# Optional

_All the changes with optional tags are not crucial enough to prevent you from getting the approval but we highly recommend to students that they should take these changes into consideration._

Cheers and Happy coding! :clap: :clap: :clap:

Feel free to leave any questions or comments in the PR thread if something is not 100% clear. Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

---------------------------------------

_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: person.rb (hunk 4 above)
Comment (karlavdelgadof) [reply person.rb]: @HAMMAS-SALEEM Thank you, I missed this accidentally, it's done :heavy_check_mark:

**Review by Lordkaito — APPROVED** [review]
Hi @karlavdelgadof, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**